intro + bfs sols for graph traversal module (#3586)
cpinitiative/usaco-guide: content/3_Silver/Graph_Traversal.mdx

</LanguageSection>

## Introduction

> The strategy of graph traversal is to visit all nodes within a graph and compute some information along the way. Two common algorithms for doing this are **DFS** and **BFS**.

Reviewer: well its not really a strategy uhm
```suggestion
The purpose of graph traversal algorithms is to visit all nodes within a graph in a certain order and compute some information along the way. Two common algorithms for doing this are **DFS** and **BFS**.
```

<LanguageSection>
<JavaSection>

```java
static final int MAX_N = (int)1e6;

public static void main(String[] args) throws IOException {
	boolean[] visited = new boolean[MAX_N];
	ArrayList<ArrayList<Integer>> adj = new ArrayList<ArrayList<Integer>>();
	for (int i = 0; i < MAX_N; i++) adj.add(new ArrayList<Integer>());

	// Define adjacency list and read in problem-specific input

	for (int i = 0; i < MAX_N; i++) {
		// iterate over all connected components in the graph
		if (!visited[i]) {
			Queue<Integer> q = new LinkedList<Integer>();
			q.add(i);

			while (q.size() > 0) {
				int currentNode = q.peek();
				q.poll();

				if (visited[currentNode]) { continue; }
				visited[currentNode] = true;

				for (int neighbor : adj.get(currentNode)) { q.add(neighbor); }
			}
		}
	}
}
```

</JavaSection>
<CPPSection>

```cpp
const int MAX_N = 1e6;

int main() {
	vector<int> adj[MAX_N];
	bool visited[MAX_N] = {0};

	// Define adjacency list and read in problem-specific input

	for (int i = 0; i < MAX_N; i++) {
		// iterate over all connected components in the graph
		if (!visited[i]) {
			queue<int> q;
			q.push(i);

			while (!q.empty()) {
				int current_node = q.front();
				q.pop();

> 				if (visited[current_node]) { continue; }

Reviewer: i suggest checking for visited nodes inside the neighbors loop, in some instances, it's a pretty significant speed build up (since youre constantly pushing already visited nodes in the queue)

				visited[current_node] = true;

				for (int neighbor : adj[current_node]) { q.push(neighbor); }
			}
		}
	}
}
```

</CPPSection>
<PySection>

```py
from collections import deque

MAX_N = int(1e6)
adj = [[] for _ in range(MAX_N)]
visited = [False] * MAX_N

# Define adjacency list and read in problem-specific input

for i in range(MAX_N):
	# iterate over all connected components in the graph
	if not visited[i]:
		q = deque([i])

		while q:
			node = q.popleft()
			visited[node] = True

			for neighbor in adj[node]:
				if not visited[neighbor]:
					q.append(neighbor)
```

</PySection>
</LanguageSection>

<Resources>
	<Resource source="CPH" title="12 - Graph traversal" starred />
</Resources>

## Applications

### Connected Components

<FocusProblem problem="connSample" />

> A **connected component** is a maximal set of connected nodes in an undirected

Reviewer: maximal?

Author: i didn't write this, but i believe that's the technical definition :/

let me know if you think this should be changed.

Reviewer: ok idc then lol

<LanguageSection>
<PySection>

```py
import sys

sys.setrecursionlimit(1000000)  # may need if input is large

MAX_N = int(1e6)
adj = [[] for _ in range(MAX_N)]
visited = [False] * MAX_N

"""
Define adjacency list and read in problem-specific input
"""

def dfs(current_node):
	visited[current_node] = True
	for neighbor in adj[current_node]:
		if not visited[neighbor]:
			dfs(neighbor)

for i in range(MAX_N):
	if not visited[i]:
		dfs(i)
```
</PySection>
<JavaSection>

```java
static ArrayList<ArrayList<Integer>> adj;
static boolean[] visited;
static final int MAX_N = (int)1e6;

public static void main(String[] args) throws IOException {
	visited = new boolean[MAX_N];
	adj = new ArrayList<ArrayList<Integer>>();
	for (int i = 0; i < MAX_N; i++) adj.add(new ArrayList<Integer>());
	/*
	 * Read in problem-specific input
	 */

	for (int i = 0; i < MAX_N; i++) {
		// iterate over all connected components in the graph
		if (!visited[i]) { dfs(i); }
	}
}

static void dfs(int currentNode) {
	if (visited[currentNode]) { return; }

	visited[currentNode] = true;
	for (int neighbor : adj.get(currentNode)) { dfs(neighbor); }
}

> ```

Reviewer 2: can you put newlines between the backticks & the html tags

</JavaSection>
<CPPSection>

```cpp
const int MAX_N = 1e6;
vector<int> adj[MAX_N];
bool visited[MAX_N];

void dfs(int current_node) {
	if (visited[current_node]) { return; }
	visited[current_node] = true;

	for (int neighbor : adj[current_node]) { dfs(neighbor); }
}

int main() {
	/*
	 * Define adjacency list and read in problem-specific input

> 	 */

Reviewer 2: make this a one-line comment

	for (int i = 0; i < MAX_N; i++) {
		// iterate over all connected components in the graph
		if (!visited[i]) { dfs(i); }
	}
}
```
</CPPSection>
</LanguageSection>

> 

Reviewer 2: ```suggestion
```

<CPPSection>

```cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

const int MN = 1e5 + 1;
int N, M, rep[MN];
vector<int> adj_list[MN];
bool visited[MN];

void bfs(int node) {
	queue<int> q;
	q.push(node);
	visited[node] = true;
	while (!q.empty()) {
		int curr = q.front();
		q.pop();
		for (int u : adj_list[curr]) {
			if (!visited[u]) {
				q.push(u);
				visited[u] = true;
			}
		}
	}
}
int count_components() {

> 	int count = 0;

Reviewer: ```suggestion
	int count = 0; // count of connected components processed
```
